File: bindings/cpp/src/pipeline.cpp
/*
 * pipeline.cpp — Triple Pipeline handle lifecycle plus the Single
 * Message cipher entries and profile registration.
 *
 * Binding-side logic is limited to the four FFI-boundary inversions:
 * caller-allocated buffers with the codified retry-once on
 * Status::BufferTooSmall, RAII handle release, byte transport, and
 * status relay. Everything else — validation, profile catalogue,
 * wire format, MAC handling — is Go's job.
 */

#include "internal.hpp"

namespace itb {

using detail::check;
using detail::fail;
using detail::ffi_bytes;
using detail::ffi_str;

/* ------------------------------------------------------------------ */
/* init / open / lifecycle                                             */
/* ------------------------------------------------------------------ */

Pipeline Pipeline::init(std::string_view profile, const Opts &opts)
{
    const std::string prof(profile);
    std::vector<std::uint8_t> blob(detail::kBlobCap);
    std::size_t blob_len = 0;
    uintptr_t handle = 0;
    int rc = ITB_Triple_Init(ffi_str(prof), ffi_str(opts.query()),
                             blob.data(), blob.size(), &blob_len, &handle);
    if (rc == static_cast<int>(Status::BufferTooSmall) && blob_len > blob.size()) {
        /* Go closed the undersized attempt; the retry re-runs Init
         * and yields a fresh session. */
        blob.resize(blob_len);
        handle = 0;
        rc = ITB_Triple_Init(ffi_str(prof), ffi_str(opts.query()),
                             blob.data(), blob.size(), &blob_len, &handle);
    }
    check(rc, "Pipeline::init");
    blob.resize(blob_len);
    blob.shrink_to_fit();
    Pipeline pipe;
    pipe.handle_ = handle;
    pipe.blob_ = std::move(blob);
    return pipe;
}

Pipeline Pipeline::open(std::string_view profile,
                        std::span<const std::byte> blob,
                        const Opts &opts,
                        std::span<const std::byte> perm_master,
                        std::span<const std::byte> wrap_master)
{
    const std::string prof(profile);
    const bool perm_given = !perm_master.empty();
    const bool wrap_given = !wrap_master.empty();
    if (perm_given != wrap_given) {
        /* Half-supplied override pair. */
        fail(static_cast<int>(Status::BadInput), "Pipeline::open");
    }
    const std::size_t masters_count = perm_given ? 2 : 0;
    uintptr_t handle = 0;
    int rc = ITB_Triple_Open(ffi_str(prof),
                             ffi_bytes(blob), blob.size(),
                             ffi_str(opts.query()),
                             ffi_bytes(perm_master), perm_master.size(),
                             ffi_bytes(wrap_master), wrap_master.size(),
                             masters_count, &handle);
    check(rc, "Pipeline::open");
    Pipeline pipe;
    pipe.handle_ = handle;
    /* The Pipeline keeps its own copy of the blob for the accessor. */
    const auto *first = reinterpret_cast<const std::uint8_t *>(blob.data());
    pipe.blob_.assign(first, first + blob.size());
    return pipe;
}

Pipeline::~Pipeline()
{
    if (handle_ != 0) {
        /* Free runs Close first Go-side (zeroing key material); the
         * status is deliberately ignored on the release path. */
        (void)ITB_Triple_Free(handle_);
        handle_ = 0;
    }
}

Pipeline::Pipeline(Pipeline &&other) noexcept
    : handle_(other.handle_), blob_(std::move(other.blob_))
{
    other.handle_ = 0;
}

Pipeline &Pipeline::operator=(Pipeline &&other) noexcept
{
    if (this != &other) {
        if (handle_ != 0) {
            (void)ITB_Triple_Free(handle_);
        }
        handle_ = other.handle_;
        blob_ = std::move(other.blob_);
        other.handle_ = 0;
    }
    return *this;
}

void Pipeline::close() noexcept
{
    if (handle_ != 0) {
        /* Idempotent Go-side; the handle stays valid until the
         * destructor releases it. */
        (void)ITB_Triple_Close(handle_);
    }
}

/* ------------------------------------------------------------------ */
/* rekey                                                               */
/* ------------------------------------------------------------------ */

void Pipeline::rekey(std::span<const std::byte> perm_master,
                     std::span<const std::byte> wrap_master)
{
    std::vector<std::uint8_t> blob(
        blob_.size() > detail::kBlobCap ? blob_.size() : detail::kBlobCap);
    std::size_t blob_len = 0;
    int rc = ITB_Triple_Rekey(handle_,
                              ffi_bytes(perm_master), perm_master.size(),
                              ffi_bytes(wrap_master), wrap_master.size(),
                              blob.data(), blob.size(), &blob_len);
    if (rc == static_cast<int>(Status::BufferTooSmall) && blob_len > blob.size()) {
        blob.resize(blob_len);
        rc = ITB_Triple_Rekey(handle_,
                              ffi_bytes(perm_master), perm_master.size(),
                              ffi_bytes(wrap_master), wrap_master.size(),
                              blob.data(), blob.size(), &blob_len);
    }
    check(rc, "Pipeline::rekey");
    blob.resize(blob_len);
    blob.shrink_to_fit();
    blob_ = std::move(blob);
}

/* ------------------------------------------------------------------ */
/* Single Message encrypt / decrypt                                    */
/* ------------------------------------------------------------------ */

namespace {

/* Shared body for the buffer-in / buffer-out cipher entries: single
 * retry-once dispatch over the caller-allocated-buffer convention. */
using CipherFn = int (*)(uintptr_t handle, void *src, size_t src_len,
                         void *out, size_t out_cap, size_t *out_len);

std::vector<std::uint8_t> cipher_call(uintptr_t handle, CipherFn fn,
                                      std::span<const std::byte> src,
                                      const char *what)
{
    std::vector<std::uint8_t> buf(detail::out_cap(src.size()));
    std::size_t n = 0;
    int rc = fn(handle, ffi_bytes(src), src.size(), buf.data(), buf.size(), &n);
    if (rc == static_cast<int>(Status::BufferTooSmall) && n > buf.size()) {
        buf.resize(n);
        rc = fn(handle, ffi_bytes(src), src.size(), buf.data(), buf.size(), &n);
    }
    check(rc, what);
    buf.resize(n);
    buf.shrink_to_fit();
    return buf;
}

} // namespace

std::vector<std::uint8_t> Pipeline::encrypt_message(std::span<const std::byte> plain) const
{
    return cipher_call(handle_, ITB_Triple_EncryptMessage, plain,
                       "Pipeline::encrypt_message");
}

std::vector<std::uint8_t> Pipeline::decrypt_message(std::span<const std::byte> wire) const
{
    return cipher_call(handle_, ITB_Triple_DecryptMessage, wire,
                       "Pipeline::decrypt_message");
}

/* ------------------------------------------------------------------ */
/* Profile registration                                                */
/* ------------------------------------------------------------------ */

void register_profile(std::string_view name, const Opts &opts)
{
    const std::string prof(name);
    check(ITB_Triple_RegisterProfile(ffi_str(prof), ffi_str(opts.query())),
          "register_profile");
}

} // namespace itb

File: bindings/zig/src/pipeline.zig
//! Triple Pipeline session over the C binding's `itb_pipeline` handle.

const std = @import("std");
const ffi = @import("ffi.zig").c;
const err = @import("error.zig");
const opts_mod = @import("opts.zig");
const stream = @import("stream.zig");

const Allocator = std.mem.Allocator;
const Opts = opts_mod.Opts;

/// Explicit master override pair for `Pipeline.load` / `loadF`. Both
/// slices must be non-empty (a half-supplied pair is rejected
/// Go-side).
pub const Masters = struct {
    perm: []const u8,
    wrap: []const u8,
};

/// A Triple Pipeline session. Every output buffer the cipher calls
/// return is owned by the caller and allocated from the `Allocator`
/// handed to `init` / `load` — release with `allocator.free`. Buffer
/// sizing and the BufferTooSmall retry-once dance live entirely in
/// the C layer; the Zig side performs no sizing of its own.
///
/// The wrapper adds no synchronisation: the Go-side Pipeline is safe
/// for concurrent cipher calls, and the Zig calls proxy straight
/// through, so the same posture holds — except `rekey` and `deinit`,
/// which must not run concurrently with cipher calls or open stream
/// sessions on the same Pipeline.
///
/// Streaming-decrypt caveat: chunked Streaming AEAD verifies per
/// chunk, so plaintext of verified chunks is released before a later
/// chunk can fail authentication.
pub const Pipeline = struct {
    allocator: Allocator,
    handle: *ffi.itb_pipeline,

    /// Constructs a fresh Pipeline against the named profile. `opts`
    /// is `null` for pure profile defaults. Release with `deinit`.
    pub fn init(allocator: Allocator, profile: [:0]const u8, opts: ?Opts) err.Error!Pipeline {
        var handle: ?*ffi.itb_pipeline = null;
        try err.check(ffi.itb_pipeline_init(profile.ptr, optsHandle(opts), &handle));
        return .{ .allocator = allocator, .handle = handle.? };
    }

    /// Reconstructs a Pipeline from a blob produced by `save` /
    /// `rekey`. `masters` is `null` to use the blob-embedded masters,
    /// or a pair to override them. The profile shape travels inside
    /// the blob — no profile name, no opts. A blob whose record names
    /// a primitive absent from the local build fails with
    /// `error.RecipePrimitiveUnknown`.
    pub fn load(allocator: Allocator, blob_bytes: []const u8, masters: ?Masters) err.Error!Pipeline {
        const m = masters orelse Masters{ .perm = &.{}, .wrap = &.{} };
        var handle: ?*ffi.itb_pipeline = null;
        try err.check(ffi.itb_pipeline_load(
            blob_bytes.ptr,
            blob_bytes.len,
            m.perm.ptr,
            m.perm.len,
            m.wrap.ptr,
            m.wrap.len,
            &handle,
        ));
        return .{ .allocator = allocator, .handle = handle.? };
    }

    /// `load` for a blob stored at `path`; the file is read inside
    /// libitb (a missing or unreadable file is `error.BadInput`).
    pub fn loadF(allocator: Allocator, path: [:0]const u8, masters: ?Masters) err.Error!Pipeline {
        const m = masters orelse Masters{ .perm = &.{}, .wrap = &.{} };
        var handle: ?*ffi.itb_pipeline = null;
        try err.check(ffi.itb_pipeline_load_f(
            path.ptr,
            m.perm.ptr,
            m.perm.len,
            m.wrap.ptr,
            m.wrap.len,
            &handle,
        ));
        return .{ .allocator = allocator, .handle = handle.? };
    }

    /// Closes (zeroing key material Go-side) and releases the handle.
    /// Call exactly once; the Pipeline must not be used afterwards,
    /// and no stream session may outlive it.
    pub fn deinit(self: *Pipeline) void {
        ffi.itb_pipeline_free(self.handle);
        self.handle = undefined;
    }

    /// The current session-bundle blob for the receiver side (the
    /// init blob, or the bytes of the latest `rekey`). Caller owns the
    /// result — release with `allocator.free`.
    pub fn save(self: *const Pipeline) err.Error![]u8 {
        var out: [*c]u8 = null;
        var out_len: usize = 0;
        const rc = ffi.itb_pipeline_save(self.handle, &out, &out_len);
        defer ffi.itb_bytes_free(out);
        try err.check(rc);
        return self.copyOut(out, out_len);
    }

    /// Writes the current blob to `path` inside libitb (mode 0600; the
    /// containing directory must exist).
    pub fn saveF(self: *const Pipeline, path: [:0]const u8) err.Error!void {
        try err.check(ffi.itb_pipeline_save_f(self.handle, path.ptr));
    }

    /// Sets the worker cap for every subsequent cipher call. `n` is
    /// clamped by libitb (`<= 0` selects auto, `> 256` becomes 256);
    /// only the handle state is reported.
    pub fn maxWorkers(self: *const Pipeline, n: i32) err.Error!void {
        try err.check(ffi.itb_pipeline_max_workers(self.handle, n));
    }

    /// Rotates the parallax + wrapper masters and returns the fresh
    /// blob (also available through `save`). Caller owns the result.
    /// Must not run concurrently with cipher calls or open stream
    /// sessions on the same Pipeline.
    pub fn rekey(self: *Pipeline, perm: []const u8, wrap: []const u8) err.Error![]u8 {
        var out: [*c]u8 = null;
        var out_len: usize = 0;
        const rc = ffi.itb_pipeline_rekey(
            self.handle,
            perm.ptr,
            perm.len,
            wrap.ptr,
            wrap.len,
            &out,
            &out_len,
        );
        defer ffi.itb_bytes_free(out);
        try err.check(rc);
        return self.copyOut(out, out_len);
    }

    /// Single Message encrypt: one call, one self-contained wire.
    /// Caller owns the result — release with `allocator.free`.
    pub fn encryptMessage(self: *const Pipeline, plain: []const u8) err.Error![]u8 {
        return self.cipher(ffi.itb_pipeline_encrypt_message, plain);
    }

    /// Receive-side counterpart of `encryptMessage`.
    pub fn decryptMessage(self: *const Pipeline, wire: []const u8) err.Error![]u8 {
        return self.cipher(ffi.itb_pipeline_decrypt_message, wire);
    }

    /// One-shot stream encrypt for callers holding the whole
    /// plaintext in memory: a single buffer-in / buffer-out FFI round
    /// trip through the Pipeline's stream chain. For bounded-memory
    /// streaming use `encryptStream` / `encryptStreamPump`. Caller
    /// owns the result — release with `allocator.free`.
    pub fn encryptStreamOneShot(self: *const Pipeline, plain: []const u8) err.Error![]u8 {
        return self.cipher(ffi.itb_pipeline_encrypt_stream_one_shot, plain);
    }

    /// Receive-side counterpart of `encryptStreamOneShot`.
    pub fn decryptStreamOneShot(self: *const Pipeline, wire: []const u8) err.Error![]u8 {
        return self.cipher(ffi.itb_pipeline_decrypt_stream_one_shot, wire);
    }

    /// Pumps the whole plaintext through an incremental encrypt
    /// session with bounded feed / drain slices C-side and returns
    /// the concatenated wire. Caller owns the result.
    pub fn encryptStreamPump(self: *const Pipeline, plain: []const u8) err.Error![]u8 {
        return self.cipher(ffi.itb_pipeline_encrypt_stream_pump, plain);
    }

    /// Receive-side counterpart of `encryptStreamPump`.
    pub fn decryptStreamPump(self: *const Pipeline, wire: []const u8) err.Error![]u8 {
        return self.cipher(ffi.itb_pipeline_decrypt_stream_pump, wire);
    }

    /// Opens an incremental encrypt session (plaintext in, wire out).
    /// Release with the session's `deinit` before this Pipeline's.
    pub fn encryptStream(self: *Pipeline) err.Error!stream.EncryptStream {
        return stream.EncryptStream.begin(self);
    }

    /// Opens an incremental decrypt session (wire in, plaintext out).
    pub fn decryptStream(self: *Pipeline) err.Error!stream.DecryptStream {
        return stream.DecryptStream.begin(self);
    }

    /// Shared body for the six buffer-in / buffer-out cipher
    /// entries: run the C call (which allocates the output and
    /// handles retry-once internally), copy into an allocator-owned
    /// slice, release the C buffer.
    fn cipher(
        self: *const Pipeline,
        func: CipherFn,
        src: []const u8,
    ) err.Error![]u8 {
        var out: [*c]u8 = null;
        var out_len: usize = 0;
        const rc = func(self.handle, src.ptr, src.len, &out, &out_len);
        defer ffi.itb_bytes_free(out); // NULL-safe; out is NULL on failure
        try err.check(rc);
        return self.copyOut(out, out_len);
    }

    fn copyOut(self: *const Pipeline, out: [*c]u8, out_len: usize) err.Error![]u8 {
        const copy = try self.allocator.alloc(u8, out_len);
        if (out_len > 0) @memcpy(copy, out[0..out_len]);
        return copy;
    }
};

const CipherFn = *const fn (
    ?*const ffi.itb_pipeline,
    [*c]const u8,
    usize,
    [*c][*c]u8,
    [*c]usize,
) callconv(.c) ffi.itb_status;

fn optsHandle(opts: ?Opts) ?*const ffi.itb_opts {
    return if (opts) |o| o.handle else null;
}
